doc/installing-packages.rst

```diff
@@ -185,6 +185,27 @@ The part of the path will be used to determine the cache key part.
     as described in the previous sections, thus requiring manual construction
     of ``01-index.tar`` file.
 
+It is possible to define ``preferred-versions``, containing additional version constraints
+for deprecating or preferring certain package versions, in the given directory.
+
+For example, if ``/absolute/path/to/directory`` looks like
+::
+
+    /absolute/path/to/directory/
+        foo-0.1.0.0.tar.gz
+        bar-0.2.0.0.tar.gz
+        preferred-versions
+
+then package deprecations and preferences will be taken into account by the solver.
+
+The contents of ``preferred-versions`` is a list of package version constraints, e.g.
+::
+
+    binary < 0.8.0.0 || > 0.8.0.0
+    text == 1.2.0.0
+
+thus, looks similar to a ``package-name.cabal``'s ``build-depends`` section.
+
```

Comment:
I'm not sure where to add this, and whether it should be mentioned at all, but this adds a new (useful?) side-effect; We can have a `file+noindex` repository with *just* `preferred-versions` and that way affect other repositories (e.g. Hackage) indices.

There is a comment *-- Note: preferences combined without using CombineStrategy*, so there is no control over that *yet*.

Reply:
Imo, it shouldn't be mentioned yet. At least some tests would be necessary before we mention it.

Reply:
But they are affecting Hackage, I'm pretty sure. If we can be sure what will be in the future, there should be a warning.

Reply:
I can't find anything obvious about merging repositories (maybe hidden or deeper in the guts?). Can you point me to some Module or something else I should take a closer look at?

Reply:
https://github.com/haskell/cabal/blob/89df281c3f74c0976f6b95524fd85c468f77ce1a/cabal-install/src/Distribution/Client/IndexUtils.hs#L301-L332

Reply:
Thanks, I will investigate!

Reply:
Took me long enough, but I can finally confirm that indeed, the `preferred-versions` file can be used to restrict the package set from Hackage. Will update the documentation.

Reply:
No idea whether it is really possible to add a test for it.